### DISCUSSION TITLE
Allow custom Components for Handle similar to Node and Edge

### DISCUSSION THREAD

**Discussion by tinybitsofcode** [xyflow/xyflow#4029] [New Features]
Hello, first of all thanks for this amazing library. I've been looking around for a while now and `react-flow` has really great design and it's also easy to understand.

From what I can see, Nodes and Edges allow for custom components to be used for sytling (and logic), but not for Handles. Customizing handles seems to only be allowed through styles - which for most use cases it's good I think.

I wanted to have a handle where it would be a checkbox using Mantine library here: https://mantine.dev/core/checkbox/, where the checkbox would change state on hover, and the color based on whether the connection is valid or not - of course, this would mean some custom logic on top of the Checkbox component, but nothing out of the ordinary.

From what I could find in this example, https://codesandbox.io/p/sandbox/throbbing-snow-n57iwm?file=%2FColorSelectorNode.js%3A68%2C47 this would be somewhat achievable, but not using out-of-the-box components.

It seems like an obvious extension path, but I'm sure there are reasons why Handle doesn't support custom components?

**Comment by moklick** [comment]
You can pass any component as a children. Does that solve your issue?

**Reply by tinybitsofcode** [reply]
Yes, that works indeed - styling the Handle to aproximate child size would give a good hover area, and setting `pointer-events` to `none` on the child component allows pointer passthrough on the Handle.

```tsx
  <Handle
    type="source"
    position={Position.Bottom}
    style={{
      minWidth: 20,
      height: 20,
      zIndex: 2,
      backgroundColor: 'transparent',
    }}
  >
    <Checkbox
      label="Socket"
      color={valid ? 'green' : 'red'}
      checked={connected}
      readOnly
      style={{
        pointerEvents: 'none',
      }}
    />
  </Handle>
  ```

I guess there isn't a way to approximate child size with absolute positioning but this is good enough for now.

**Reply by tinybitsofcode** [reply]
Working a bit more on this, it seems quite easy to change the `source` Handle state, because we get the `onConnect` and `isValidConnection` callbacks, but would it be possible to update a `target` Handle as well?

Looking through the examples, the only one I could find that changes the target Handle state of the `target` is the [Validation](https://reactflow.dev/examples/interaction/validation) one, but the state changes internally to the library - and it's reflected through CSS styling only (the `valid` class being added to the `div`).

I guess what I'm asking is for a sort of first-citizen support for Handles too, similar to how we have `useNodes()` or `useEdges()` - it would be helpful if I could somehow update the `handle` state on events or callbacks, similar to how we do it for edges or nodes.

In my case, I wanted to mark the `target` checkbox as ticked if the connection is done. I guess firing a `onConnect` on the target handler could do.

I'm new to the library, so again, it might be something obvious that I'm missing. :)

**Comment by tinybitsofcode** [comment]
It seems the new features of v12, for computing flows are pretty much what I was looking for. `useHandleConnections`, `useNodesData` and `useConnection` are very helpful so thanks for adding them.

I'm trying to use the `useConnection` hook to update Handle color based on the connection details. 

I see the `endHandle` property is missing if the connection is `invalid`, but exists when the connection is `valid`. I think it would be helpful to have the `endHandle` populated even if the connection is `invalid`, removing the need to do other check (e.g. check also if the mouse is hovering over a certain Handle). 

Would that be possible to populate `endHandle` on `invalid` since I guess the object is already there?
